include/nbl/builtin/hlsl/sampling/cos_weighted.hlsl

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_HLSL_SAMPLING_COS_WEIGHTED_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SAMPLING_COS_WEIGHTED_INCLUDED_
+
+#include "nbl/builtin/hlsl/concepts.hlsl"
+#include "nbl/builtin/hlsl/sampling/concentric_mapping.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+
+namespace impl
+{
+template<typename T>
+struct ProjectedHemisphere
+{
+    using vector_t2 = vector<T, 2>;
+    using vector_t3 = vector<T, 3>;
+    
+    static vector_t3 generate(vector_t2 _sample)
+    {
+        vector_t2 p = concentricMapping<T>(_sample * 0.99999f + 0.000005f);
+        T z = sqrt<T>(max(0.0, 1.0 - p.x * p.x - p.y * p.y));
+        return vector_t3(p.x, p.y, z);
+    }
+
+    static T pdf(T L_z)
+    {
+        return L_z * numbers::inv_pi<float>;
+    }
+
+    static T quotient_and_pdf(NBL_REF_ARG(T) _pdf, T L)
+    {
+        _pdf = pdf(L);
+        return 1.0;
+    }
+};
+}
+
+// probably could've split into separate structs and ProjectedSphere inherit ProjectedHemisphere
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+vector<T, 3> projected_hemisphere_generate(vector<T, 2> _sample)
```

Comment:
Yes, but use "Composition over inheritance" so the ProjectedSphere should have a hemisphere member